shared/testutil/block.go

```diff
@@ -0,0 +1,381 @@
+	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/bls"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+// BlockGenConfig is used to define the requested conditions
+// for block generation.
+type BlockGenConfig struct {
+	MaxProposerSlashings uint64
+	MaxAttesterSlashings uint64
+	MaxAttestations      uint64
+	MaxDeposits          uint64
+	MaxVoluntaryExits    uint64
+}
+
+// GenerateFullBlock generates a fully valid block with the requested parameters.
+// Use BlockGenConfig to declare the conditions you would like the block generated under.
+func GenerateFullBlock(
+	t testing.TB,
+	bState *pb.BeaconState,
+	privs []*bls.SecretKey,
+	conf *BlockGenConfig,
+) *ethpb.BeaconBlock {
+	currentSlot := bState.Slot
+
+	pSlashings := []*ethpb.ProposerSlashing{}
+	if conf.MaxProposerSlashings > 0 {
+		pSlashings = generateProposerSlashings(t, bState, privs, conf.MaxProposerSlashings)
+	}
+
+	aSlashings := []*ethpb.AttesterSlashing{}
+	if conf.MaxAttesterSlashings > 0 {
+		aSlashings = generateAttesterSlashings(t, bState, privs, conf.MaxAttesterSlashings)
+	}
+
+	atts := []*ethpb.Attestation{}
+	if conf.MaxAttestations > 0 {
+		atts = generateAttestations(t, bState, privs, conf.MaxAttestations)
+	}
+
+	newDeposits, eth1Data := []*ethpb.Deposit{}, &ethpb.Eth1Data{}
+	if conf.MaxDeposits > 0 {
+		newDeposits, eth1Data = generateDepositsAndEth1Data(t, bState, conf.MaxDeposits)
+	}
+
+	exits := []*ethpb.VoluntaryExit{}
+	if conf.MaxVoluntaryExits > 0 {
+		exits = generateVoluntaryExits(t, bState, privs, conf.MaxVoluntaryExits)
+	}
+
+	newHeader := proto.Clone(bState.LatestBlockHeader).(*ethpb.BeaconBlockHeader)
+	prevStateRoot, err := ssz.HashTreeRoot(bState)
+	if err != nil {
+		t.Fatal(err)
+	}
+	newHeader.StateRoot = prevStateRoot[:]
+	parentRoot, err := ssz.SigningRoot(newHeader)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	bState.Slot++
+	reveal, err := CreateRandaoReveal(bState, helpers.CurrentEpoch(bState), privs)
+	if err != nil {
+		t.Fatal(err)
+	}
+	bState.Slot--
+
+	block := &ethpb.BeaconBlock{
+		Slot:       currentSlot + 1,
+		ParentRoot: parentRoot[:],
+		Body: &ethpb.BeaconBlockBody{
+			Eth1Data:          eth1Data,
+			RandaoReveal:      reveal,
+			ProposerSlashings: pSlashings,
+			AttesterSlashings: aSlashings,
+			Attestations:      atts,
+			VoluntaryExits:    exits,
+			Deposits:          newDeposits,
+		},
+	}
+
+	s, err := state.CalculateStateRoot(context.Background(), bState, block)
+	if err != nil {
+		t.Fatal(err)
+	}
+	root, err := ssz.HashTreeRoot(s)
+	if err != nil {
+		t.Fatal(err)
+	}
+	block.StateRoot = root[:]
+	blockRoot, err := ssz.SigningRoot(block)
+	if err != nil {
+		t.Fatal(err)
+	}
+	bState.Slot++
+	proposerIdx, err := helpers.BeaconProposerIndex(bState)
+	if err != nil {
+		t.Fatal(err)
+	}
+	bState.Slot--
+	domain := helpers.Domain(bState.Fork, helpers.CurrentEpoch(bState), params.BeaconConfig().DomainBeaconProposer)
```

Comment:
What is this `bState.Slot++` and then `bState.Slot--` back and forth? 😅

Reply:
This is because the block slot is one slot ahead of the beacon state slot, and since the functions used for beacon proposer index are deterministic to the beacon state slot, we need to increment it to what it should be and use that.

Reply:
Yeah. Can you leave a comment for it? Thanks